source/common/router/rds_impl.cc

```diff
@@ -227,37 +227,37 @@ RouteConfigProviderManagerImpl::RouteConfigProviderManagerImpl(Server::Admin& ad
   RELEASE_ASSERT(config_tracker_entry_, "");
 }
 
-Router::RouteConfigProviderPtr RouteConfigProviderManagerImpl::createRdsRouteConfigProvider(
+Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::createRdsRouteConfigProvider(
     const envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
     Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
     Init::Manager& init_manager) {
   // RdsRouteConfigSubscriptions are unique based on their serialized RDS config.
   const uint64_t manager_identifier = MessageUtil::hash(rds);
-  auto& server_factory_context = factory_context.getServerFactoryContext();
+  auto it = dynamic_route_config_providers_.find(manager_identifier);
 
-  RdsRouteConfigSubscriptionSharedPtr subscription;
-
-  auto it = route_config_subscriptions_.find(manager_identifier);
-  if (it == route_config_subscriptions_.end()) {
+  if (it == dynamic_route_config_providers_.end()) {
     // std::make_shared does not work for classes with private constructors. There are ways
     // around it. However, since this is not a performance critical path we err on the side
     // of simplicity.
-    subscription.reset(new RdsRouteConfigSubscription(
-        rds, manager_identifier, server_factory_context, factory_context.messageValidationVisitor(),
-        init_manager, stat_prefix, *this));
+    RdsRouteConfigSubscriptionSharedPtr subscription(new RdsRouteConfigSubscription(
+        rds, manager_identifier, factory_context.getServerFactoryContext(),
+        factory_context.messageValidationVisitor(), factory_context.initManager(), stat_prefix,
+        *this));
     init_manager.add(subscription->init_target_);
-    route_config_subscriptions_.insert({manager_identifier, subscription});
+    std::shared_ptr<RdsRouteConfigProviderImpl> new_provider{
+        new RdsRouteConfigProviderImpl(std::move(subscription), factory_context)};
+    dynamic_route_config_providers_.insert(
+        {manager_identifier, std::weak_ptr<RdsRouteConfigProviderImpl>(new_provider)});
+    return new_provider;
   } else {
     // Because the RouteConfigProviderManager's weak_ptrs only get cleaned up
     // in the RdsRouteConfigSubscription destructor, and the single threaded nature
     // of this code, locking the weak_ptr will not fail.
-    subscription = it->second.lock();
+    auto existing_provider = it->second.lock();
+    RELEASE_ASSERT(existing_provider != nullptr, "");
```

Comment:
This should be a regular `ASSERT` I think.

Reply:
1 line below the existing_provider is the deref. Instead of a nullptr deref I prefer to throw a message here. I added a detailed error message in the follow commit.
There is no penalty if the condition is true.